pkg/parse/run.go

```diff
@@ -81,54 +81,77 @@ func Run(ctx context.Context, p Parser) {
+			// Reset the cache partially to make sure all the steps of a parse-apply-watch loop will run.
 			// The cached sourceState will not be reset to avoid reading all the source files unnecessarily.
-			state.resetAllButSourceState()
+			// The cached needToRetry will not be reset to avoid resetting the backoff retries.
+			state.resetPartialCache()
 			run(ctx, p, triggerResync, state)
 
 			resyncTimer.Reset(opts.resyncPeriod)             // Schedule resync attempt
 			retryTimer.Reset(opts.retryPeriod)               // Schedule retry attempt
 			statusUpdateTimer.Reset(opts.statusUpdatePeriod) // Schedule status update attempt
 
 		// Re-import declared resources from the filesystem (from git-sync).
+		// If the reconciler is in the process of reconciling a given commit, the re-import won't
+		// happen until the ongoing reconciliation is done.
 		case <-runTimer.C:
 			run(ctx, p, triggerReimport, state)
 
-			runTimer.Reset(opts.pollingPeriod)               // Schedule re-run attempt
+			runTimer.Reset(opts.pollingPeriod)               // Schedule re-import attempt
 			retryTimer.Reset(opts.retryPeriod)               // Schedule retry attempt
 			statusUpdateTimer.Reset(opts.statusUpdatePeriod) // Schedule status update attempt
 
 		// Retry if there was an error, conflict, or any watches need to be updated.
 		case <-retryTimer.C:
 			var trigger string
 			if opts.managementConflict() {
-				// Reset the cache to make sure all the steps of a parse-apply-watch loop will run.
+				// Reset the cache partially to make sure all the steps of a parse-apply-watch loop will run.
 				// The cached sourceState will not be reset to avoid reading all the source files unnecessarily.
-				state.resetAllButSourceState()
+				// The cached needToRetry will not be reset to avoid resetting the backoff retries.
+				state.resetPartialCache()
 				trigger = triggerManagementConflict
-				// When conflict is detected, wait longer (same as the polling frequency) for the next retry.
-				time.Sleep(opts.pollingPeriod)
-			} else if state.cache.needToRetry && state.cache.readyToRetry() {
-				klog.Infof("The last reconciliation failed")
+			} else if state.cache.needToRetry {
 				trigger = triggerRetry
 			} else if opts.needToUpdateWatch() {
-				klog.Infof("Some watches need to be updated")
 				trigger = triggerWatchUpdate
 			} else {
 				// Don't reset the retry timer if there's nothing to retry.
 				continue
 			}
-			run(ctx, p, trigger, state)
+			if state.backoff.Steps == 0 {
+				klog.Infof("Retry limit (%v) has been reached", retryLimit)
+				// Don't reset retryTimer if retry limit has been reached.
+				continue
+			}
+			// retryTimer is reset to `opts.retryPeriod` every time after `run` is executed
+			// to catch triggerManagementConflict/triggerRetry/triggerWatchUpdate in time.
+			// `state.NextRetryTime` is used to check when the next retry should start.
+			if !time.Now().Before(state.nextRetryTime) {
```

Comment:
The time check here, along with the timer reset, seems to use two different strategies to manage the retry backoff. Can we do `retryTimer.Reset(retryDuration)`, and get rid of the hard-coded retryPeriod?

Reply:
Unfortunately, we cannot. I added a comment here L141-143 to explain why we need to reset `retryTimer` to `opts.retryPeriod` every time after `run` . Let me know if it is clear enough.

Reply:
You are right. I have been going back and force on the experiments, and was a little overwhelmed.  Adding https://github.com/GoogleContainerTools/kpt-config-sync/pull/812/files#diff-4e49e2db3ba5fd76214acd7bc50a14b99623650c72ca6ffaf51157b2fe1f4f15R130 could avoid introducing `nextRetryTime`.